### PR TITLE
Auto generate setoutput method and use it from custom code

### PR DESCRIPTION
**Pull request by kumargauravsharma**
Issue #, if available:
`provideUpdatedResource` method at [replication_group/custom_update_api.go#L701-L866](https://github.com/aws-controllers-k8s/elasticache-controller/blob/main/pkg/resource/replication_group/custom_update_api.go#L701-L866) gets outof sync if ReplicationGroup struct changes. This method contains code which is copied from auto generated blocks inside `sdk.go`. 
Thus, as and when code generator produces changes inside `sdk.go` set output methods, the `provideUpdatedResource` method goes out of sync and requires manual copy of the changes to it.

Auto generating this function will resolve this issue.

Description of changes:
This PR is dependents upon following code-generator enhancements PRs:
1. [Enable crd variables and controller functions and hooks within hook templates](https://github.com/aws-controllers-k8s/code-generator/pull/91)
2. [Add sdk addon hook and CRD method to get UnwrappedOutputShapeName](https://github.com/aws-controllers-k8s/code-generator/pull/92)

This PR introduces a dynamic hook template which us used by code generator to auto generate the set output method for `ReplicationGroup` shape.
Old function is removed from code and references and unit tests have been updated.

Testing
`make test` passed/

Dependency: Changes in this PR are dependent on following code-generator PR: https://github.com/aws-controllers-k8s/code-generator/pull/94

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: templates/hooks/sdk_addon.go.tpl [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+		return nil, nil
+	}
+	resp := &svcsdk.{{ .CRD.Ops.Create.OutputRef.Shape.ShapeName }}{ {{ $outputShape.ShapeName }}:obj }
+	// Merge in the information we read from the API call above to the copy of
+    // the original Kubernetes object we passed to the function
+    ko := r.ko.DeepCopy()
+{{ $createCode := GoCodeSetCreateOutput .CRD "resp" "ko" 1 true }}
+{{ $createCode }}
+	rm.setStatusDefaults(ko)
+{{- if $hookCode := Hook .CRD "sdk_addon_set_output_post_populate" }}
```
Comment (A-Hilaly) [inline templates/hooks/sdk_addon.go.tpl:20]: Nit: you can also add this hook to list of known hooks (comment) in https://github.com/aws-controllers-k8s/code-generator/pull/94 - https://github.com/aws-controllers-k8s/code-generator/blob/85436ab1f0b7994e7c6651a03177ab90a8b61ad5/pkg/generate/ack/hook.go#L60

**Review by A-Hilaly — COMMENTED** [review]
Nice! 👍 

if you regen using `v0.2.3` you will also get a `ack-generate-metadata.yaml`

File: templates/hooks/sdk_addon.go.tpl (hunk 1 above)
Comment (kumargauravsharma) [reply templates/hooks/sdk_addon.go.tpl:20]: As this dynamic template and the hook id that it consumes, is scoped to the service controller and is specified in generator config for elasticache controller.
This way any dynamic template can open its extension points by supporting such hooks.

As this hook is specific to this service controller scoped template I considered it local.

File: templates/hooks/sdk_addon.go.tpl [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+) (*resource, error) {
+	if obj == nil ||
+		r == nil ||
+		r.ko == nil {
+		return nil, nil
+	}
+	resp := &svcsdk.{{ .CRD.Ops.Create.OutputRef.Shape.ShapeName }}{ {{ $outputShape.ShapeName }}:obj }
+	// Merge in the information we read from the API call above to the copy of
+    // the original Kubernetes object we passed to the function
+    ko := r.ko.DeepCopy()
```
Comment (jaypipes) [inline templates/hooks/sdk_addon.go.tpl:16]: note you've got a couple indentation/spacing issues in this template. `go fmt` is executed on the resulting file, which is why these issues are corrected during `make build-controller`, but you may want to address them here anyway

Comment (kumargauravsharma) [reply templates/hooks/sdk_addon.go.tpl]: Thanks @jaypipes I have updated the template  and regenerated the code after rebasing `code-generator` and `elasticache-controller` repositories.

`make test` passed.

**Review by jaypipes — APPROVED** [review]
awesomesauce. nice work @kumargauravsharma :)

**Comment by kumargauravsharma** [comment]
e2e tests have passed and PR is approved.

**Pull request merged**